src/devices/machine/w5100.cpp

```diff
@@ -0,0 +1,4320 @@
+	, device_network_interface(mconfig, *this, 10)
+	, device_memory_interface(mconfig, *this)
+	, m_device_type(device_type)
+	, m_irq_handler(*this)
+	, m_address_space_config("memory", ENDIANNESS_BIG, 8, 15, 0, address_map_constructor())
+{
+}
+
+void w5100_base_device::device_start()
+{
+	save_item(m_common_registers, "General Registers");
+	save_item(m_socket_registers[0], "Socket 0 Registers");
+	save_item(m_socket_registers[1], "Socket 1 Registers");
+	save_item(m_socket_registers[2], "Socket 2 Registers");
+	save_item(m_socket_registers[3], "Socket 3 Registers");
+	save_item(m_tx_buffer, "TX Buffer Memory");
+	save_item(m_rx_buffer, "RX Buffer Memory");
+
+	save_item(NAME(m_idm));
+	save_item(NAME(m_identification));
+	save_item(NAME(m_irq_state));
+
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_una), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_nxt), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, rcv_nxt), 4);
+
+	save_pointer(STRUCT_MEMBER(m_tcp, irs), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, iss), 4);
+
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_wl1), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_wl2), 4);
+
+	save_pointer(STRUCT_MEMBER(m_tcp, resend_seq), 4);
+
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_wnd), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, rcv_wnd), 4);
+
+
+	m_interrupt_timer = timer_alloc(FUNC(w5100_base_device::interrupt_timer), this);
+	for (int sn = 0; sn < 4 ; ++sn)
+	{
+		emu_timer *t;
+		t = timer_alloc(FUNC(w5100_base_device::delayed_ack_timer), this);
+		t->set_param(sn);
+		m_delayed_ack_timers[sn] = t;
+
+		t = timer_alloc(FUNC(w5100_base_device::keep_alive_timer), this);
+		t->set_param(sn);
+		m_keep_alive_timers[sn] = t;
+	}
+
+	for (int sn = 0; sn < 5; ++sn)
+	{
+		m_retry_timers[sn] = timer_alloc(FUNC(w5100_base_device::retry_timer), this);
+	}
+
+	m_address_space = &space(AS_PROGRAM);
+
+	m_address_space->install_rom(0x0000, 0x00ff, 0x3f00, m_common_registers);
+	m_address_space->install_write_handler(0x0000, 0x00ff, 0, 0x3f00, 0, write8sm_delegate(*this, FUNC(w5100_base_device::write_common_register)));
+
+	if (m_device_type == dev_type::W5100S)
+	{
+		// m_address_space->install_rom(0x0400, 0x07ff, 0x3800, m_socket_registers);
+		m_address_space->install_read_handler(0x0400, 0x07ff, 0, 0x3800, 0, read8sm_delegate(*this, FUNC(w5100_base_device::read_socket_register)));
+		m_address_space->install_write_handler(0x0400, 0x07ff, 0, 0x3800, 0, write8sm_delegate(*this, FUNC(w5100_base_device::write_socket_register)));
+	}
+	else
+	{
+		// on the w5100, socket 3 is mirrored.
+		m_address_space->install_read_handler(0x0400, 0x04ff, 0x00ff, 0x3800, 0, read8sm_delegate(*this, FUNC(w5100_base_device::read_socket_register_3)));
+		m_address_space->install_write_handler(0x0400, 0x04ff, 0x00ff, 0x3800, 0, write8sm_delegate(*this, FUNC(w5100_base_device::write_socket_register_3)));
+
+		m_address_space->install_read_handler(0x0400, 0x07ff, 0, 0x0000, 0, read8sm_delegate(*this, FUNC(w5100_base_device::read_socket_register)));
```

Comment:
Why aren’t you supplying an address map function when configuring the address space as this is all static and installed on start.